File: R/priors.R
# Prior distributions for parameters — sampling and log-density.
# R twin of python/dssatcalibrator/priors.py. Seeded from
# DSSAT_Calibration/priors.R (regularization / shrinkage idiom).
#
# Supported distributions (all truncated to [min, max]):
#   uniform (default) | normal | lognormal | triangular
#
# Parity note: log-density functions are deterministic and are checked exactly
# against the Python implementation. Sampling uses base-R RNG streams, which
# cannot match NumPy bit-for-bit, so sampling parity is verified statistically
# (matching moments) per the agreed policy.

.prior_bounds <- function(spec) c(as.numeric(spec$min), as.numeric(spec$max))

.prior_center <- function(spec) {
  b <- .prior_bounds(spec); lo <- b[1]; hi <- b[2]
  prior <- spec$prior
  if (is.null(prior)) prior <- list()
  if (!is.null(prior$mean)) return(as.numeric(prior$mean))
  if (!is.null(prior$mode)) return(as.numeric(prior$mode))
  if (!is.null(spec$start)) return(as.numeric(spec$start))
  0.5 * (lo + hi)
}

.prior_dist_name <- function(spec) {
  prior <- spec$prior
  if (is.null(prior) || is.null(prior$dist)) return("uniform")
  tolower(as.character(prior$dist))
}

# Inverse-CDF draw from a triangular(lo, mode, hi).
.rtriangular <- function(n, lo, mode, hi) {
  u <- runif(n)
  fc <- if (hi > lo) (mode - lo) / (hi - lo) else 0.5
  out <- numeric(n)
  left <- u < fc
  out[left]  <- lo + sqrt(u[left] * (hi - lo) * (mode - lo))
  out[!left] <- hi - sqrt((1 - u[!left]) * (hi - lo) * (hi - mode))
  out
}

#' Draw `n` samples for one parameter from its (truncated) prior.
#' Mirrors priors.py:sample_one. `rng` is accepted for signature parity and
#' ignored (base-R RNG is used; seed via set.seed()).
#' @export
sample_one <- function(spec, n, rng = NULL) {
  b <- .prior_bounds(spec); lo <- b[1]; hi <- b[2]
  dist <- .prior_dist_name(spec)
  prior <- spec$prior; if (is.null(prior)) prior <- list()

  if (dist == "uniform") {
    return(runif(n, lo, hi))
  }
  if (dist == "normal") {
    mu <- .prior_center(spec)
    sd <- if (!is.null(prior$sd)) as.numeric(prior$sd) else 0.25 * (hi - lo)
    # Truncated normal via inverse CDF on [lo, hi].
    plo <- pnorm(lo, mu, sd); phi <- pnorm(hi, mu, sd)
    u <- runif(n, plo, phi)
    return(qnorm(u, mu, sd))
  }
  if (dist == "lognormal") {
    center <- max(.prior_center(spec), 1e-9)
    sigma <- if (!is.null(prior$sd)) as.numeric(prior$sd) else 0.5
    draws <- rlnorm(n, meanlog = log(center), sdlog = sigma)
    return(pmin(pmax(draws, lo), hi))
  }
  if (dist == "triangular") {
    mode <- min(max(.prior_center(spec), lo), hi)
    return(.rtriangular(n, lo, mode, hi))
  }
  stop(sprintf("unknown prior dist '%s' for parameter '%s'", dist, spec$name))
}

#' Log prior density of one parameter value (-Inf outside the bounds).
#' Mirrors priors.py:log_prior_one (and matches scipy's truncnorm/lognorm/triang
#' parameterisations).
#' @export
log_prior_one <- function(spec, value) {
  b <- .prior_bounds(spec); lo <- b[1]; hi <- b[2]
  if (!(lo <= value && value <= hi)) return(-Inf)
  dist <- .prior_dist_name(spec)
  prior <- spec$prior; if (is.null(prior)) prior <- list()

  if (dist == "uniform") return(0.0)

  if (dist == "normal") {
    mu <- .prior_center(spec)
    sd <- if (!is.null(prior$sd)) as.numeric(prior$sd) else 0.25 * (hi - lo)
    xi <- (value - mu) / sd
    logZ <- log(pnorm(hi, mu, sd) - pnorm(lo, mu, sd))
    # scipy truncnorm.logpdf = norm.logpdf(xi) - log(scale) - log(Z)
    return(dnorm(xi, log = TRUE) - log(sd) - logZ)
  }
  if (dist == "lognormal") {
    center <- max(.prior_center(spec), 1e-9)
    sigma <- if (!is.null(prior$sd)) as.numeric(prior$sd) else 0.5
    return(dlnorm(value, meanlog = log(center), sdlog = sigma, log = TRUE))
  }
  if (dist == "triangular") {
    mode <- min(max(.prior_center(spec), lo), hi)
    width <- hi - lo
    if (width <= 0) return(-Inf)
    if (value <= mode) {
      dens <- if (mode > lo) 2 * (value - lo) / (width * (mode - lo)) else 0
    } else {
      dens <- if (hi > mode) 2 * (hi - value) / (width * (hi - mode)) else 0
    }
    return(if (dens > 0) log(dens) else -Inf)
  }
  stop(sprintf("unknown prior dist '%s' for parameter '%s'", dist, spec$name))
}

#' Draw `n` parameter sets, one column per active parameter.
#' Returns a data.frame in native units. Mirrors priors.py:sample_prior_design.
#' @export
sample_prior_design <- function(space, n, rng = NULL) {
  cols <- lapply(space$specs, function(s) sample_one(s, n, rng))
  names(cols) <- space$names
  as.data.frame(cols, check.names = FALSE)
}

#' Total log prior of a parameter vector = sum over independent parameters.
#' Mirrors priors.py:log_prior_vec.
#' @export
log_prior_vec <- function(space, theta) {
  total <- 0.0
  for (s in space$specs) {
    total <- total + log_prior_one(s, as.numeric(theta[[s$name]]))
    if (is.infinite(total) && total < 0) break
  }
  total
}

#' TRUE if any active parameter declares a non-uniform prior.
#' Mirrors priors.py:has_informative_prior.
#' @export
has_informative_prior <- function(space) {
  any(vapply(space$specs, function(s) .prior_dist_name(s) != "uniform", logical(1)))
}

File: python/dssatcalibrator/priors.py
"""Prior distributions for parameters — sampling and log-density.

Why this module exists
----------------------
The config lets every parameter declare a ``prior`` (see ``CONCEPT.md`` §4)::

    P5: { active: true, min: 300, max: 700, start: 505, prior: {dist: normal, sd: 50} }

A *prior* is your belief about a parameter **before** looking at the data: a
plausible value (``start``), a spread (``sd``), and hard bounds (``min``/``max``).
The Bayesian engines (GLUE, SMC-PF, MCMC) combine this prior with how well each
parameter set fits the observations to produce a *posterior* (your belief
**after** seeing the data).

Earlier versions of the framework silently ignored the ``prior`` field and
always assumed a flat (uniform) prior between ``min`` and ``max``. This module
makes the declared prior actually count, while keeping ``uniform`` as the safe
default when no ``prior`` is given.

Supported distributions (all truncated to ``[min, max]``):

======================  ==========================================================
``dist``                meaning
======================  ==========================================================
``uniform`` (default)   every value in ``[min, max]`` equally likely
``normal``              bell curve centred on ``start`` (or ``prior.mean``),
                        width ``prior.sd``; values outside the bounds are excluded
``lognormal``           skewed, positive-only; ``prior.sd`` is the sd of log(x)
``triangular``          peak at ``start`` (or ``prior.mode``), zero at the bounds
======================  ==========================================================

Everything is deliberately small and dependency-light: it uses only SciPy
(already a core dependency). To add a new distribution, add one ``elif`` branch
in both :func:`sample_one` and :func:`log_prior_one` — nothing else changes.
"""

from __future__ import annotations

import numpy as np
from scipy import stats


def _bounds(spec: dict) -> tuple[float, float]:
    return float(spec["min"]), float(spec["max"])


def _center(spec: dict) -> float:
    """The prior's central value: explicit ``prior.mean``/``mode`` else ``start``
    else the midpoint of the range."""
    lo, hi = _bounds(spec)
    prior = spec.get("prior") or {}
    if "mean" in prior:
        return float(prior["mean"])
    if "mode" in prior:
        return float(prior["mode"])
    if "start" in spec and spec["start"] is not None:
        return float(spec["start"])
    return 0.5 * (lo + hi)


def _dist_name(spec: dict) -> str:
    prior = spec.get("prior") or {}
    return str(prior.get("dist", "uniform")).lower()


def sample_one(spec: dict, n: int, rng: np.random.Generator) -> np.ndarray:
    """Draw ``n`` samples for a single parameter from its (truncated) prior."""
    lo, hi = _bounds(spec)
    dist = _dist_name(spec)
    prior = spec.get("prior") or {}

    if dist == "uniform":
        return rng.uniform(lo, hi, size=n)

    if dist == "normal":
        mu = _center(spec)
        sd = float(prior.get("sd", 0.25 * (hi - lo)))
        # Truncated normal: a, b are the bounds expressed in standard deviations.
        a, b = (lo - mu) / sd, (hi - mu) / sd
        return stats.truncnorm.rvs(a, b, loc=mu, scale=sd, size=n, random_state=rng)

    if dist == "lognormal":
        # prior.sd is the sd of log(x); centre is the median (exp of log-mean).
        center = max(_center(spec), 1e-9)
        sigma = float(prior.get("sd", 0.5))
        draws = rng.lognormal(mean=np.log(center), sigma=sigma, size=n)
        return np.clip(draws, lo, hi)

    if dist == "triangular":
        mode = _center(spec)
        return rng.triangular(lo, np.clip(mode, lo, hi), hi, size=n)

    raise ValueError(f"unknown prior dist '{dist}' for parameter '{spec.get('name')}'")


def log_prior_one(spec: dict, value: float) -> float:
    """Log prior density of one parameter value (``-inf`` outside the bounds).

    Normalising constants are dropped where they cancel in Metropolis-Hastings
    ratios; what matters is the *shape* and that out-of-bounds values are
    impossible.
    """
    lo, hi = _bounds(spec)
    if not (lo <= value <= hi):
        return float("-inf")
    dist = _dist_name(spec)
    prior = spec.get("prior") or {}

    if dist == "uniform":
        return 0.0  # constant inside the box

    if dist == "normal":
        mu = _center(spec)
        sd = float(prior.get("sd", 0.25 * (hi - lo)))
        a, b = (lo - mu) / sd, (hi - mu) / sd
        return float(stats.truncnorm.logpdf(value, a, b, loc=mu, scale=sd))

    if dist == "lognormal":
        center = max(_center(spec), 1e-9)
        sigma = float(prior.get("sd", 0.5))
        return float(stats.lognorm.logpdf(value, s=sigma, scale=center))

    if dist == "triangular":
        mode = float(np.clip(_center(spec), lo, hi))
        c = (mode - lo) / (hi - lo) if hi > lo else 0.5
        return float(stats.triang.logpdf(value, c, loc=lo, scale=(hi - lo)))

    raise ValueError(f"unknown prior dist '{dist}' for parameter '{spec.get('name')}'")


def sample_prior_design(space, n: int, rng: np.random.Generator):
    """Draw ``n`` parameter sets, one column per active parameter.

    Returns a ``pandas.DataFrame`` in native parameter units (same shape the
    samplers in :mod:`samplers` produce), so engines can use prior draws or
    space-filling draws interchangeably.
    """
    import pandas as pd

    cols = {s["name"]: sample_one(s, n, rng) for s in space.specs}
    return pd.DataFrame(cols, columns=space.names)


def log_prior_vec(space, theta: dict) -> float:
    """Total log prior of a parameter vector = sum over independent parameters."""
    total = 0.0
    for s in space.specs:
        total += log_prior_one(s, float(theta[s["name"]]))
        if total == float("-inf"):
            break
    return total


def has_informative_prior(space) -> bool:
    """True if *any* active parameter declares a non-uniform prior.

    Lets engines skip the prior term entirely (it is a no-op constant) when every
    parameter is uniform — keeping the old, simpler behaviour bit-for-bit.
    """
    return any(_dist_name(s) != "uniform" for s in space.specs)
